vault/external_tests/identity/aliases_test.go

```diff
@@ -250,3 +255,881 @@ func TestIdentityStore_RenameAlias_CannotMergeEntity(t *testing.T) {
+	}
+
+	// Now create an entity and alias for bob
+	entityRespBob, err := client.Logical().Write("identity/entity", map[string]interface{}{
+		"name": "bob-smith",
+	})
+	if err != nil {
+		t.Fatalf("err:%v resp:%#v", err, entityRespBob)
+	}
+	if entityRespBob == nil {
+		t.Fatalf("expected a non-nil response")
+	}
+	entityIdBob := entityRespBob.Data["id"].(string)
+	if entityIdBob == "" {
+		t.Fatal("EntityID not present in response")
+	}
+
+	aliasResp, err := client.Logical().Write("identity/entity-alias", map[string]interface{}{
+		"name":           "bob",
+		"canonical_id":   entityIdBob,
+		"mount_accessor": mountAccessor,
+	})
+	if err != nil {
+		t.Fatalf("err:%v resp:%#v", err, aliasResp)
+	}
+
+	aliasIdBob := aliasResp.Data["id"].(string)
+	if aliasIdBob == "" {
+		t.Fatal("Alias ID not present in response")
+	}
+
+	// Create userpass login for alice
+	_, err = client.Logical().Write("auth/userpass/users/alice", map[string]interface{}{
+		"password": "training",
+	})
+	if err != nil {
+		t.Fatal(err)
+	}
+	_, err = client.Logical().Write("auth/userpass/login/alice", map[string]interface{}{
+		"password": "training",
+	})
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	// Now create a new unrelated entity and alias, alice-smith
+	entityRespAlice, err := client.Logical().Write("identity/entity", map[string]interface{}{
+		"name": "alice-smith",
+	})
+	if err != nil {
+		t.Fatalf("err:%v resp:%#v", err, entityRespAlice)
+	}
+	if entityRespAlice == nil {
+		t.Fatalf("expected a non-nil response")
+	}
+	entityIdAlice := entityRespAlice.Data["id"].(string)
+	if entityIdAlice == "" {
+		t.Fatal("EntityID not present in response")
+	}
+
+	aliasResp, err = client.Logical().Write("identity/entity-alias", map[string]interface{}{
+		"name":           "alice",
+		"canonical_id":   entityIdAlice,
+		"mount_accessor": mountAccessor,
+	})
+	if err != nil {
+		t.Fatalf("err:%v resp:%#v", err, aliasResp)
+	}
+
+	// Try and login with alias 2 (alice) pre-merge
+	loginResp, err := client.Logical().Write("auth/userpass/login/alice", map[string]interface{}{
+		"password": "training",
+	})
+	if err != nil {
+		t.Fatalf("err:%v resp:%#v", err, loginResp)
+	}
+
+	// Perform entity merge
+	mergeResp, err := client.Logical().Write("identity/entity/merge", map[string]interface{}{
+		"to_entity_id":                  entityIdBob,
+		"from_entity_ids":               entityIdAlice,
+		"conflicting_alias_ids_to_keep": aliasIdBob,
+	})
+	if err != nil {
+		t.Fatalf("err:%v resp:%#v", err, mergeResp)
+	}
+
+	// Delete entity id 1 (bob)
+	deleteResp, err := client.Logical().Delete(fmt.Sprintf("identity/entity/id/%s", entityIdBob))
+	if err != nil {
+		t.Fatalf("err:%v resp:%#v", err, deleteResp)
+	}
+
+	// Try and login with alias 2 (alice) post-merge
+	userpassAuth, err := auth.NewUserpassAuth("alice", &auth.Password{FromString: "training"})
+	if err != nil {
+		t.Fatal(err)
+	}
+	loginResp, err = client.Auth().Login(context.Background(), userpassAuth)
```

Comment:
I think the entityID is returned by the login request, so you could validate that it remains unchanged, as opposed to the original entity having been deleted in the merge and recreated by the new login.  Also, why use this login method here and a Logical.Write earlier?

Reply:
The reason for the difference is that the Login method actually logs in (i.e. the client is modified by the request, so we shouldn't do it earlier). I'm happy to unify the approaches though. Good suggestion for the entityId validation - that has been added.

Reply:
Ah, you mean it sets the token on the client?  Gotcha, well you might want to add a comment to explain this, no biggie though.